design/general-progress-monitoring.md

```diff
@@ -0,0 +1,496 @@
+reconciliation, backups which do not have a `velero-backup.json` object in the object store will be
+ignored.
+
+## Plug-in API changes
+
+### OperationProgress struct
+
+    type OperationProgress struct {
+        completed bool                          // True when the operation has completed, either successfully or with a failure
+        err error                               // Set when the operation has failed
+        itemsCompleted, itemsToComplete int64   // The number of items that have been completed and the items to complete
+                                                // For a disk, an item would be a byte and itemsToComplete would be the
+                                                // total size to transfer (may be less than the size of a volume if
+                                                // performing an incremental) and itemsCompleted is the number of bytes
+                                                // transferred.  On successful completion, itemsCompleted and itemsToComplete
+                                                // should be the same
+        started, updated time.Time              // When the upload was started and when the last update was seen.  Not all
+                                                // systems retain when the upload was begun, return Time 0 (time.Unix(0, 0))
+                                                // if unknown.
+    }
+
+### VolumeSnapshotter changes
+
+Two new methods will be added to the VolumeSnapshotter interface:
+
+    Progress(snapshotID string) (UploadProgress, error)
+    Cancel(snapshotID string) (UploadProgress, error)
+
+Open question: Does VolumeSnapshotter need Cancel, or is that only needed for BIA/RIA?
+
+Progress will report the current status of a snapshot upload.  This should be callable at
+any time after the snapshot has been taken.  In the event a plug-in is restarted, if the operationID
+(snapshot ID) continues to be valid it should be possible to retrieve the progress.
+
+`error` is set if there is an issue retrieving progress.  If the snapshot is has encountered an
+error during the upload, the error should be returned in OperationProgress and error should be nil.
+
+### BackupItemAction and RestoreItemAction plug-in changes
+
+Currently CSI snapshots and the Velero Plug-in for vSphere are implemented as BackupItemAction
+plugins.  While the majority of BackupItemAction plugins do not take snapshots or upload data, this
+functionality is useful for any longstanding plugin operation managed by an external
+process/controller so we will modify BackupItemAction and RestoreItemAction to optionally return an
+operationID in addition to the modified item.
+
+The Backup/RestoreItemAction plugin identifier as well as the Item and Snapshot ID will be stored in
+the `<backup-name>-itemoperations.json.gz`.  When checking for progress, this info will be used to
```

Comment:
So the operationID will be the snapshotID for volumesnapshotter plugins. For data mover BIA/RIA plugins, it will be an appropriate ID for the datamover CRs. The item should also just be an identifier, not the entire resource. Adding an example to the doc.